src/App.tsx

```diff
@@ -1,26 +1,186 @@
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState(Filter.ALL);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const setError = (message: string) => {
+    setErrorMessage(message);
+
+    setTimeout(() => {
+      setErrorMessage('');
+    }, 3000);
+  };
+
+  useEffect(() => {
+    setIsLoading(true);
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setError('Unable to load todos');
+      })
+      .finally(() => setIsLoading(false));
+  }, []);
+
+  const isAllCompleted = useMemo(() => {
+    return todos.every(el => el.completed);
+  }, [todos]);
+
+  const isAllIncompleted = useMemo(() => {
+    return todos.every(el => !el.completed);
+  }, [todos]);
+
+  const todosForView = useMemo(() => {
+    switch (filter) {
+      case Filter.ALL:
+        return todos;
+      case Filter.ACTIVE:
+        return todos.filter(el => !el.completed);
+      case Filter.COMPLETED:
+        return todos.filter(el => el.completed);
+      default:
+        return todos;
+    }
+  }, [todos, filter]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const addNewTodo = (title: string) => {
+    return new Promise((resolve, reject) => {
+      const tempTodo: Todo = {
+        title: title,
+        userId: USER_ID,
+        completed: false,
+        id: 0,
+      };
+
+      setTodos(currentTodos => [...currentTodos, tempTodo]);
+
+      addTodo({
+        title: title,
+        userId: USER_ID,
+        completed: false,
+      })
+        .then(todo => {
+          setTodos(currentTodos => [
+            ...currentTodos.filter(el => el.id !== 0),
+            todo,
+          ]);
+          resolve(true);
+        })
+        .catch(() => {
+          setError('Unable to add a todo');
+          setTodos(currentTodos => [...currentTodos.filter(el => el.id !== 0)]);
+          reject(false);
+        })
+        .finally(() => {
+          setTimeout(() => {
+            inputRef.current?.focus();
+          }, 0);
+        });
+    });
+  };
+
+  const handleCompleteTodo = (id: number) => {
+    const newTodoList = todos.map(el => {
+      return el.id !== id ? el : { ...el, completed: !el.completed };
+    });
+
+    setTodos(newTodoList);
+  };
+
+  const handleDeleteTodo = (id: number) => {
+    return new Promise((resolve, reject) => {
+      deleteTodo(id)
+        .then(() => {
+          setTodos(currentTodos => [
+            ...currentTodos.filter(el => el.id !== id),
+          ]);
+          resolve(true);
+        })
+        .catch(() => {
+          setError('Unable to delete a todo');
+          setTodos(currentTodos => [...currentTodos.filter(el => el.id !== 0)]);
+          reject(false);
+        })
+        .finally(() => {
+          setTimeout(() => {
+            inputRef.current?.focus();
+          }, 0);
+        });
+    });
+  };
```

Comment:
```suggestion
    return deleteTodo(id)
        .then(() => {
          setTodos(currentTodos => [
            ...currentTodos.filter(el => el.id !== id),
          ]);
        })
        .catch(() => {
          setError('Unable to delete a todo');
          setTodos(currentTodos => [...currentTodos.filter(el => el.id !== 0)]);
        })
        .finally(() => {
          setTimeout(() => {
            inputRef.current?.focus();
          }, 0);
        });
    });
  };
```
you don't need the new promise for adding and deleting todos